physical/spanner/spanner_ha.go

```diff
@@ -0,0 +1,409 @@
+
+	r, err := l.get(context.Background())
+	if err != nil {
+		return false, "", err
+	}
+	if r == nil {
+		return false, "", err
+	}
+	return true, string(r.Value), nil
+}
+
+// attemptLock attempts to acquire a lock. If the given channel is closed, the
+// acquisition attempt stops. This function returns when a lock is acquired or
+// an error occurs.
+func (l *Lock) attemptLock(stopCh <-chan struct{}) (bool, error) {
+	ticker := time.NewTicker(l.retryInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			acquired, err := l.writeLock()
+			if err != nil {
+				return false, errors.Wrap(err, "attempt lock")
+			}
+			if !acquired {
+				continue
+			}
+
+			return true, nil
+		case <-stopCh:
+			return false, nil
+		}
+	}
+}
+
+// renewLock renews the given lock until the channel is closed.
+func (l *Lock) renewLock() {
+	ticker := time.NewTicker(l.renewInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			l.writeLock()
+		case <-l.stopCh:
+			return
+		}
+	}
+}
+
+// watchLock checks whether the lock has changed in the table and closes the
+// leader channel accordingly. If an error occurs during the check, watchLock
+// will retry the operation and then close the leader channel if it can't
+// succeed after retries.
+func (l *Lock) watchLock() {
+	retries := 0
+	ticker := time.NewTicker(l.watchRetryInterval)
+
+	for {
+		// Check if the channel is already closed
+		select {
+		case <-l.stopCh:
+		default:
+		}
+
+		// Check if we've exceeded retries
+		if retries >= l.watchRetryMax-1 {
+			break
+		}
+
+		// Wait for the timer
+		<-ticker.C
+
+		// Attempt to read the key
+		r, err := l.get(context.Background())
+		if err != nil {
+			retries++
+			continue
+		}
+
+		// Verify the identity is the same
+		if r == nil || r.Identity != l.identity {
+			break
+		}
+	}
+
+	l.stopLock.Lock()
+	defer l.stopLock.Unlock()
+	if !l.stopped {
+		l.stopped = true
+		close(l.stopCh)
+	}
+}
+
+// writeLock writes the given lock using the following algorith:
+//
+// - lock does not exist
+//   - write the lock
+// - lock exists
+//   - if key is empty or identity is the same or timestamp exceeds TTL
+//     - update the lock to self
+func (l *Lock) writeLock() (bool, error) {
+	// Pooling
+	l.backend.permitPool.Acquire()
+	defer l.backend.permitPool.Release()
+
+	// Keep track of whether the lock was written
+	lockWritten := false
+
+	// Create a transaction to read and the update (maybe)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// The transaction will be retried, and it could sit in a queue behind, say,
+	// the delete operation. To stop the transaction, we close the context when
+	// the associated stopCh is received.
+	go func() {
+		select {
+		case <-l.stopCh:
+			cancel()
+		case <-ctx.Done():
+		}
+	}()
+
+	_, err := l.backend.client.ReadWriteTransaction(ctx, func(ctx context.Context, txn *spanner.ReadWriteTransaction) error {
```

Comment:
This is pretty cool; I'm assuming that the entire block will happen atomically so you can't get two clients elevating themselves.

Reply:
Correct. The function within the function may be called multiple times, but the transaction is retained during retries. It's a global lock basically).